python/graphstorm/gpartition/random_partition.py

```diff
@@ -0,0 +1,137 @@
+"""
+    Copyright 2023 Contributors
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+    Random partition algorithm
+"""
+import os
+import logging
+import json
+
+import numpy as np
+import pyarrow as pa
+import pyarrow.csv as pa_csv
+
+class RandomPartitionAlgorithm():
```

Comment:
Do we plan to implement more single-instance partition algorithms? (e.g. range)

If yes, it's worth creating a class hierarchy with a shared interface, i.e. a `PartitionAlgorithm` base class from which `RandomPartitionAlgorithm` inherits, where we define a public `create_partitions(args.num_parts, local_output_path)` method, that's implemented as

```
def create_partitions(self, args.num_parts: int, local_output_path: str) -> str:
    partition_dir = os.path.join(self.local_output_path, "partition_assignment")
    os.makedirs(partition_dir, exist_ok=True)
    self._run_partitioning(args.num_parts, local_output_path)
    self._create_metadata(args.num_parts, local_output_path)

    return partition_dir
```

If we do that then we can probably remove `LocalRandomPartitioner` and only have a hierarchy of `PartitionAlgorithm` classes (any common init code can go to the base `PartitionAlgorithm` class)
that implement the two "private" functions, `_run_partitioning` and `_create_metadata`.

So in `python/graphstorm/gpartition/dist_partition_graph.py` we'd have:


```
if args.part_algorithm == "random":
    partitioner = RandomPartitionAlgorithm(metadata_file)
elif args.part_algorithm == "range":
    partitioner = RangePartitionAlgorithm(metadata_file)
else:
    raise NotImplementedError()

local_partition_path = partitioner.create_partitions(args.num_parts, local_output_path)

[...]
```

Note that we move `local_output_path` from object state to function argument (the same partitioner should be able to output partitions anywhere, so let's remove it from the object's state).

Reply:
SageMakerPartitioner also calls RandomPartitionAlgorithm.run_partitioning.
If we change the API here, we need to change the SageMakerPartitioner.

Reply:
Agreed, right now we have

```python
        random_part = RandomPartitionAlgorithm(self.local_output_path,
                                               self.metadata,
                                               self.rank)
        partition_dir = random_part.run_partitioning(num_partitions)
```

in `SageMakerPartitioner()._run_partitioning(num_partitions: int)`. What I'm suggesting would change that to:


```python
        random_part = RandomPartitionAlgorithm(
                                               self.metadata,
                                               self.rank)
        partition_dir = random_part.create_partitions(num_partitions, self.local_output_path)
```

Does this make sense?